pycbc/derived_parameters.py

```diff
@@ -0,0 +1,462 @@
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+"""
+This modules provides a library of functions that calculate waveform parameters
+from other parameters. All exposed functions in this module's namespace return
+one parameter given a set of inputs.
+"""
+
+import copy
+import numpy
+import lal
+from pycbc import coordinates
+from pycbc.detector import Detector
+
+#
+# =============================================================================
+#
+#                           Helper functions
+#
+# =============================================================================
+#
+def _ensurearray(arg):
+    """Ensures that the given argument is a numpy array. If it is not, the
+    argument is converted to an array.
+    """
+    if not isinstance(arg, numpy.ndarray):
+        arr = numpy.array(arg)
+        arg = arr
+    return arg
+
+
+def _formatreturn(arg):
+    """If the given argument is a numpy array with shape (1,), just returns
+    that value."""
+    if arg.size == 1:
+        arg = arg.item()
+    return arg
+
+#
+# =============================================================================
+#
+#                           CBC mass functions
+#
+# =============================================================================
+#
+def primary_mass(mass1, mass2):
+    """Returns the larger of mass1 and mass2 (p = primary)."""
+    mass1 = _ensurearray(mass1)
+    mass2 = _ensurearray(mass2)
+    if mass1.shape != mass2.shape:
+        raise ValueError("mass1 and mass2 must have same shape")
+    mp = copy.copy(mass1)
+    mask = mass1 < mass2
+    mp[mask] = mass2[mask]
+    return _formatreturn(mp)
+
+
+def secondary_mass(mass1, mass2):
+    """Returns the smaller of mass1 and mass2 (s = secondary)."""
+    mass1 = _ensurearray(mass1)
+    mass2 = _ensurearray(mass2)
+    if mass1.shape != mass2.shape:
+        raise ValueError("mass1 and mass2 must have same shape")
+    ms = copy.copy(mass2)
+    mask = mass1 < mass2
+    ms[mask] = mass1[mask]
+    return _formatreturn(ms)
+
+
+def mtotal_from_mass1_mass2(mass1, mass2):
```

Comment:
Because I usually find myself in the position of wanting a particular parameter and looking for what function I can use to get it, rather than having some set of parameters and wondering what parameters I can derive using them. For example, if you want `mchirp`, you can just do `derived_parameters.mchirp_` + tab complete in ipython to see what functions will let you get that parameter. I never liked the opposite convention used in pnutils for that reason.